malariagen_data/ag3.py

```diff
@@ -341,3 +347,100 @@ def _results_cache_add_analysis_params(self, params):
+
+    def load_inversion_tags(self, inversion: str):
+        # needs to be modified depending on where we are hosting
+        path = "malariagen_data/resources/karyotype_tag_snps.csv"
+        df_tag_snps = pd.read_csv(path, sep=",")
+        return df_tag_snps.query(f"inversion == '{inversion}'").reset_index()
+
+    def karyotype_tags_n_alt(self, gt, alts, inversion_alts):
+        # could be Numba'd for speed but was already quick (not many inversion tag snps)
+        n_sites = gt.shape[0]
+        n_samples = gt.shape[1]
+
+        # create empty array
+        inv_n_alt = np.empty((n_sites, n_samples), dtype=np.int8)
+
+        # for every site
+        for i in range(n_sites):
+            # find the index of the correct tag snp allele
+            tagsnp_index = np.where(alts[i] == inversion_alts[i])[0]
+
+            for j in range(n_samples):
+                # count alleles which == tag snp allele and store
+                n_tag_alleles = np.sum(gt[i, j] == tagsnp_index[0])
+                inv_n_alt[i, j] = n_tag_alleles
+
+        return inv_n_alt
+
+    @check_types
+    def karyotype(
+        self,
+        inversion,
+        sample_sets: Optional[base_params.sample_sets] = None,
+        sample_query: Optional[base_params.sample_query] = None,
+    ) -> pd.DataFrame:
+        """
+        Infer karyotype from tag SNPs for a given inversion.
+
+        Parameters
+        ----------
+        inversion : str
+            Name of inversion to infer karyotype for. One of ["2La", "2Rb", "2Rc_gam", "2Rc_col", "2Rd", "2Rj"].
+        sample_sets : list of str, optional
+            Sample sets to include in the analysis. If None (default), use all sample sets.
+        sample_query : str, optional
+            Query to filter samples. If None (default), use all samples.
+        """
+
+        inversion_list = ["2La", "2Rb", "2Rc_gam", "2Rc_col", "2Rd", "2Rj"]
+        assert (
+            inversion in inversion_list
+        ), f"Inversion {inversion} not found - please select one of {inversion_list}"
+
+        # load tag snp data
+        df_tagsnps = self.load_inversion_tags(inversion=inversion)
+        inversion_pos = df_tagsnps["position"]
+        inversion_alts = df_tagsnps["alt_allele"]
+        contig = inversion[0:2]
+
+        # get snp calls for inversion region
+        start, end = np.min(inversion_pos), np.max(inversion_pos)
+        region = f"{contig}:{start}-{end}"
+
+        ds_snps = self.snp_calls(
+            region=region, sample_sets=sample_sets, sample_query=sample_query
+        )
+        geno = allel.GenotypeDaskArray(ds_snps["call_genotype"].data)
+        pos = allel.SortedIndex(ds_snps["variant_position"].values)
+        samples = ds_snps["sample_id"].values
+        alts = ds_snps["variant_allele"].values.astype(str)
+
+        # subset to position of inversion tags
+        mask = pos.locate_intersection(inversion_pos)[0]
+        alts = alts[mask]
+        geno = geno.compress(mask, axis=0)
```

Comment:
```suggestion
        geno = geno.compress(mask, axis=0).compute()
```

For simplicity, suggest to compute this into memory here.